Add ref for nn.functional.prelu

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79768

TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

soulitzer committed:
Add ref for nn.functional.prelu

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20262,6 +20262,9 @@ def __init__(
             DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_executor'),
         ),
     ),
+    PythonRefInfo(
```
Cool! Follow-up question on testing:

- should we make prelu (and this reference info) elementwise unary? It's a little tricky because prelu takes a weight parameter, but other elementwise unary operations take additional kwargs, too, and we probably have the test architecture to handle it
- whether we make prelu elementwise unary or not, are the additional prelu-specific inputs sufficient to validate the weight parameter is working properly? Should we add more? Maybe even adding a "reference inputs" function for prelu with extended samples?
- does prelu have ErrorInputs for its custom checks, like channel-related checks? If not, should we add them?

soulitzer replied:
Ah, I will think about these points more!

soulitzer replied:
Updated to make prelu elementwise unary, weight is always scalar in sample kwargs, but the 1-D case is tested in other sample inputs. Also added ErrorInputs.

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -376,3 +378,37 @@ def gelu(a: TensorLikeType, approximate: str = "none") -> TensorLikeType:
         raise RuntimeError("approximate argument must be either none or tanh.")
+
+
+def prelu(a: TensorLikeType, weight: TensorLikeType) -> TensorLikeType:
```
Nice!

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -376,3 +378,37 @@ def gelu(a: TensorLikeType, approximate: str = "none") -> TensorLikeType:
+    )
+
+    if weight.numel() != 1:
+        check(a.ndim > 0, lambda: "Not allow zero-dim input tensor.")
```
Great use of existing helper functionality

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -376,3 +378,37 @@ def gelu(a: TensorLikeType, approximate: str = "none") -> TensorLikeType:
+        lambda: f"prelu: Expected `weight` to be a scalar or 1D tensor, but got: "
+        f"ndim = {weight.ndim}",
+    )
+    weight = prims.broadcast_in_dim(
```
Instead of calling the broadcast_in_dim prim directly (which looks totally correct, by the way), should we instead call broadcast_to here? 

https://github.com/pytorch/pytorch/blob/459090e3ceaff7276a4fbf51161e3a8e627a275f/torch/_refs/__init__.py#L1680

Generally we want to prefer calling the most specific reference implementation for the behavior we're interested in.

soulitzer replied:
I see. Not sure if `broadast_to` works when `a.ndim > 2` though:
```
>>> import torch
>>> a = torch.ones(3, 2, 3)
>>> b = torch.ones(2)
>>> torch.prelu(a, b)
tensor([[[1., 1., 1.],
         [1., 1., 1.]],

        [[1., 1., 1.],
         [1., 1., 1.]],

        [[1., 1., 1.],
         [1., 1., 1.]]])
>>> torch.broadcast_to(b, (3, 2, 3))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The expanded size of the tensor (3) must match the existing size (2) at non-singleton dimension 2.  Target sizes: [3, 2, 3].  Tensor sizes: [2]
>>> 
```

ngimel replied:
Wait, why does prelu(a,b) work - b is not broadcastable to a, it should error out.

mruberry replied:
Bonus points if we can make the docs more readable, @soulitzer

soulitzer committed:
Update on "Add ref for nn.functional.prelu"




[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

jbschlosser commented on torch/nn/functional.py:
```diff
@@ -1609,6 +1609,14 @@ def leaky_relu(input: Tensor, negative_slope: float = 0.01, inplace: bool = Fals
 learnable parameter.
 
+.. note::
+    `weight` is expected to be a scalar or 1-D tensor. If `weight` is 1-D,
+    it must have the same size as the size of `input`'s channel dim, usually the
+    2nd dim of input (or when input has dims < 2 the numbers of channels = 1).
+    In the 1-D case, note that when `input` has dim > 2, `weight` can be expanded
+    to the shape of `input` in a way that is not possible using normal
+    :ref:`broadcasting semantics<broadcasting-semantics>`.
```
```suggestion
    `weight` is expected to be a scalar or 1-D tensor. If `weight` is 1-D,
    its size must match the number of input channels, determined by
    `input.size(1)` when `input.dim() >= 2`, otherwise 1.
    In the 1-D case, note that when `input` has dim > 2, `weight` can be expanded
    to the shape of `input` in a way that is not possible using normal
    :ref:`broadcasting semantics<broadcasting-semantics>`.
```

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer commented:
Since NumPy does not support bfloat16, in the sample kwargs I defaulted to use float32 when dtypes is bfloat16. I also needed to skip the test_python_ref tests for bfloat16 because they actually check if the dtypes match. Maybe there is a better way to go about this though.

mruberry commented:
> Since NumPy does not support bfloat16, in the sample kwargs I defaulted to use float32 for weight when dtypes is bfloat16.

OK -- let's just add a comment

> I also needed to skip the test_python_ref tests for bfloat16 because they actually check if the dtypes match (test reference numerics does not care). Maybe there is a better way to go about this though.

Shouldn't test_python_ref also not care and just perform type promotion?

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -376,3 +378,37 @@ def gelu(a: TensorLikeType, approximate: str = "none") -> TensorLikeType:
         raise RuntimeError("approximate argument must be either none or tanh.")
+
+
+def prelu(a: TensorLikeType, weight: TensorLikeType) -> TensorLikeType:
```
I think this needs a type promotion wrapper? That should resolve the ref's discrepancy with eager when `a` and `weight` have distinct dtypes

soulitzer replied:
Shouldn't it already have type promotion from `mul` and `where`? Would it still benefit from the wrapper here?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1935,6 +1935,32 @@ def sample_inputs_nn_functional_prelu(op_info, device, dtype, requires_grad, **k
+    return ({'weight': weight}, {'weight': np_weight})
+
+def error_inputs_prelu(op, device):
+    # Weight has numel != 1, but self.ndim is zero-dim tensor
```
Nice error inputs and comments for each case

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1935,6 +1935,32 @@ def sample_inputs_nn_functional_prelu(op_info, device, dtype, requires_grad, **k
     yield SampleInput(make_arg((S, S)), kwargs=dict(weight=weight_tensor,))
     yield SampleInput(make_arg((S, S)), kwargs=dict(weight=make_arg((S,)),))
 
+def sample_kwargs_prelu_scalar_weight(device, dtype, input):
```
Add a comment for what's up with NumPy -- can't the NumPy weight be float32 but the torch weight still in bfloat16?

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -14404,6 +14430,9 @@ def error_inputs_mean(op_info, device, **kwargs):
         supports_gradgrad=True,
         supports_out=False,
+        # test_reference_numerics only tests the case when the weight tensor is a scalar
+        sample_kwargs=sample_kwargs_prelu_scalar_weight,
```
The sample inputs should also yield from the elementwise unary sample inputs, and a custom reference inputs should be added that does the same mutatis mutandi (yielding from the elementwise unary reference inputs)

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20262,6 +20291,17 @@ def __init__(
+        torch_opinfo_name="nn.functional.prelu",
+        skips=(
+            # NumPy does not support bfloat16
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref',
```
But it's test_numpy_ref that doesn't support bfloat16 -- I think this is just a type promotion discrepancy

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -20262,6 +20291,17 @@ def __init__(
+            # NumPy does not support bfloat16
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref',
+                         dtypes=(torch.bfloat16,)),
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_torch_fallback',
```
These were fixed after adding the type promotion wrapper

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

mruberry commented:
Looks great except the one job failure looks relevant:

```
======================================================================
FAIL [0.012s]: test_errors_nn_functional_prelu_cuda (__main__.TestCommonCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 1806, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 377, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 786, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 821, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 979, in only_fn
    return fn(self, *args, **kwargs)
  File "test_ops.py", line 543, in test_errors
    op(si.input, *si.args, **si.kwargs)
AssertionError: RuntimeError not raised
----------------------------------------------------------------------
```

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

mruberry reviewed (approved):
Cool! Ship it!

soulitzer commented:
@pytorchbot rebase

soulitzer committed:
Update on "Add ref for nn.functional.prelu"


TODO:
- not sure if these error-inputs work for all devices (awaiting CI)

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
